### PR TITLE
James 2195

### PR DESCRIPTION
**Pull request by strokyl**
So please only review last three commits !

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 1]
```diff
@@ -244,7 +245,7 @@ public void appendMultipartMessageWithOneAttachments(String username, String mes
 
     @Given("^the user has a multipart message \"([^\"]*)\" in \"([^\"]*)\" mailbox$")
     public void appendMultipartMessageWithOneAttachments(String messageName, String arg1) throws Exception {
-        appendMessage(messageName, "eml/htmlAndTextMultipartWithOneAttachment.eml");
+        appendMessage(messageName, arg1, "eml/htmlAndTextMultipartWithOneAttachment.eml");
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: s/arg1/mailbox

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 2]
```diff
@@ -558,9 +563,9 @@ public void assertIdOfTheFirstMessage(String messageName) throws Exception {
     }
 
     @Then("^the message is in \"([^\"]*)\" mailboxes")
-    public void assertMailboxIdsOfTheFirstMessage(String mailboxIds) throws Exception {
+    public void assertMailboxIdsOfTheFirstMessage(String mailboxNames) throws Exception {
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: s/MailboxIds/MailboxNames/

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 3]
```diff
@@ -569,7 +574,25 @@ public void assertMailboxIdsOfTheFirstMessage(String mailboxIds) throws Exceptio
             .distinct()
             .collect(Guavate.toImmutableList());
         assertThat(jsonPath.<JSONArray>read(FIRST_MESSAGE + ".mailboxIds"))
-            .hasSize(2)
+            .hasSize(values.size())
+            .containsOnlyElementsOf(values);
+    }
+
+    @Then("^the message is in \"([^\"]*)\" user mailboxes")
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: the difference between these to then is tricky

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java [hunk 4]
```diff
@@ -99,40 +115,112 @@ public void copyMessageToMailbox(String message, String sourceMailbox, String de
...
+            "    \"#0\"" +
+            "  ]" +
+            "]");
         mainStepdefs.awaitMethod.run();
     }
+
+    @Then("^the message \"([^\"]*)\" is not updated$")
+    public void assertIdOfTheFirstMessage(String messageName) throws Exception {
+        MessageId id = getMessagesMethodStepdefs.getMessageId(messageName);
+        System.out.println(jsonPath);
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java]: debug

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#                                                              *
+# Unless required by applicable law or agreed to in writing,   *
+# software distributed under the License is distributed on an  *
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+# KIND, either express or implied.  See the License for the    *
+# specific language governing permissions and limitations      *
+# under the License.                                           *
+# **************************************************************/
+
+Feature: GetMessages method
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: SetMessages

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  As a James user
+  I want to be able to retrieve my messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: not exactly

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 7]
```diff
@@ -569,7 +574,25 @@ public void assertMailboxIdsOfTheFirstMessage(String mailboxIds) throws Exceptio
...
+    public void assertMailboxIdsOfTheFirstMessageWithUser(String mailboxIds) throws Exception {
+        List<String> values = Splitter.on(",")
+            .splitToList(mailboxIds).stream()
+            .map(name -> Splitter.on(':').omitEmptyStrings().splitToList(name))
+            .peek(nameParts -> Preconditions.checkArgument(nameParts.size() == 2))
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: I improved this part using DataTable

**Comment by strokyl** [comment]
```
CassandraSetMessagesMethodTest>SetMessagesMethodTest.messageWithNonIndexableAttachmentShouldHaveItsEmailBodyIndexed:3811 1 expectation failed.
JSON path [0][1].messageIds doesn't match.
Expected: a collection containing "8b0c97d0-b557-11e7-be56-7d198c646607"
  Actual: [8b21f490-b557-11e7-be56-7d198c646607]
```

**Comment by rouazana** [comment]
please be careful when you rebase, you lost my fixup commits!

**Comment by rouazana** [comment]
restored

**Comment by strokyl** [comment]
test this please

**Comment by chibenwa** [comment]
```
[f458427658a0619d8d082747060c80f96f54f6ed] [INFO] Results:
[f458427658a0619d8d082747060c80f96f54f6ed] [INFO] 
[f458427658a0619d8d082747060c80f96f54f6ed] [ERROR] Errors: 
[f458427658a0619d8d082747060c80f96f54f6ed] [ERROR]   ReadOnlyUsersLDAPRepositoryTest.unknownUserShouldNotBeAbleToLogInWithVirtualHosting:159->startUsersRepository:63 » Communication
[f458427658a0619d8d082747060c80f96f54f6ed] [INFO] 
[f458427658a0619d8d082747060c80f96f54f6ed] [ERROR] Tests run: 11, Failures: 0, Errors: 1, Skipped: 0
``` 

test this please

**Comment by strokyl** [comment]
test this please

**Comment by chibenwa** [comment]
Again fail to launch unit tests. I will re-open this.

**Pull request closed without merging**